esp-bootloader-esp-idf/src/ota.rs

```diff
@@ -0,0 +1,316 @@
+//! # Over The Air Updates (OTA)
+//!
+//! ## Overview
+//! The OTA update mechanism allows a device to update itself based on data
+//! received while the normal firmware is running (for example, over Wi-Fi,
+//! Bluetooth or Ethernet).
+//!
+//! OTA requires configuring the Partition Tables of the device with at least
+//! two OTA app slot partitions (i.e., ota_0 and ota_1) and an OTA Data
+//! Partition.
+//!
+//! The OTA operation functions write a new app firmware image to whichever OTA
+//! app slot that is currently not selected for booting. Once the image is
+//! verified, the OTA Data partition is updated to specify that this image
+//! should be used for the next boot.
+//!
+//! Note: The prebuilt bootloaders provided by `espflash` _might not_ include
+//! OTA support. In that case you need to built the bootloader yourself.
+//!
+//! The general procedure to change the active slot
+//! - read the partition table [crate::partitions::read_partition_table]
+//! - find the Data/Ota partition
+//! - initialize [Ota]
+//! - read the current slot, change the current slot
+//!
+//! For more details see <https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/ota.html>
+use crc::{Algorithm, Crc};
+use embedded_storage::{ReadStorage, Storage};
+
+use crate::partitions::FlashRegion;
+
+#[link_section = ".espressif.metadata"]
+#[used]
+#[export_name = "bootloader.FEATURE_OTA"]
+static OTA_FEATURE: [u8; 11] = *b"FEATURE=OTA";
+
+static ALGO: Algorithm<u32> = Algorithm {
+    width: 32,
+    poly: 0x04c11db7,
+    init: 0,
+    refin: true,
+    refout: true,
+    xorout: 0xffffffff,
+    check: 0,
+    residue: 0,
+};
+
+const SLOT0_DATA_OFFSET: u32 = 0x0000;
+const SLOT1_DATA_OFFSET: u32 = 0x1000;
```

Comment:
There can be up to 16 OTA partitions.

Reply:
yes and no ..... bootloader will never deal with more than two ota partitions - you can have more but the layout of the ota-data partition is fixed (to my knowledge)